spec.bs

```diff
@@ -718,35 +780,81 @@ To <dfn>create a request</dfn> given a [=URL=] |url|, a [=string=] <dfn for="cre
   1. If [=header list/getting a structured field value=] "X-Allow-Protected-Audience" from |headers| does not
-    return true, set |resource| to failure and return.
+    return true, return false.
   1. Let |headerMimeType| be the result of [=header list/extracting a MIME type=] from |headers|.
-  1. Set |resource| to failure and return, if any of the following conditions hold:
+  1. Return false if any of the following conditions hold:
     * |headerMimeType| is failure;
     * |mimeType| is "`text/javascript`" and |headerMimeType| is not a [=JavaScript MIME type=];
     * |mimeType| is "`application/json`" and |headerMimeType| is not a [=JSON MIME type=].
   1. Let |mimeTypeCharset| be |headerMimeType|'s [=MIME type/parameters=]["`charset`"].
-  1. Set |resource| to failure and return if any of the following conditions hold:
+  1. Return false if any of the following conditions hold:
     * If |mimeTypeCharset| does not [=map/exist=], or |mimeTypeCharset| is "utf-8", and
       |responseBody| is not [=UTF-8=] encoded;
     * If |mimeTypeCharset| is "us-ascii", and |responseBody| is not [=ascii string=].
-  1. Set |resource| to |responseBody|.
-1. Wait for |resource| to be set.
-1. Return |resource|.
+  1. Return true.
+</div>
 
+<div algorithm>
+To <dfn>fetch script</dfn> given a [=URL=] |url|:
+
+  1. Let |request| be the result of [=creating a request=] with |url|, "`text/javascript`", and
+    null.
+  1. Let |script| be null.
+  1. [=Fetch=] |request| with [=fetch/processResponseConsumeBody=] set to the following steps given
+    a [=response=] |response| and |responseBody|:
+    1. If [=validate fetching response=] with |response|, |responseBody| and "`text/javascript`"
+      returns false, set |script| to failure and return.
+    1. Set |script| to |responseBody|.
+  1. Wait for |script| to be set.
+  1. Return |script|.
+</div>
+
+<div algorithm>
+To <dfn>fetch trusted signals</dfn> given a [=URL=] |url|, and a [=boolean=] |isBiddingSignal|:
+
+  1. Let |request| be the result of [=creating a request=] with |url|, "`application/json`", and
+    null.
+  1. Let |signals| be null.
+  1. Let |dataVersion| be null.
+  1. Let |formatVersionString| be null.
+  1. [=Fetch=] |request| with [=fetch/processResponseConsumeBody=] set to the following steps given
+    a [=response=] |response| and |responseBody|:
+    1. If [=validate fetching response=] with |response|, |responseBody| and "`application/json`"
+      returns false, set |signals| to failure and return.
+    1. Let |headers| be |response|'s [=response/header list=].
+    1. Let |dataVersionString| be the result of [=header list/getting a structured field value=]
+      "Data-Version" from |headers|.
+    1. If |dataVersionString| is not null:
+      1. If |dataVersionString| [=string/starts with=] "`0`" but [=string/is=] not "`0`", set
+        |signals| to failure and return.
+      1. Let |dataVersion| be the result of parsing |dataVersionString| to a {{unsigned long}}.

```

Comment:
This is a bit odd... if Data-Version a structured header, I think we can just get the right type? https://www.rfc-editor.org/rfc/rfc8941.html#name-integers (I guess its up to whatever is setting the value to be an unsigned long).

Reply:
I think we need to define the header as one consisting of certain sub type, like https://wicg.github.io/nav-speculation/speculation-rules.html#speculation-rules-header. So basically a list whose values are certain types, or a simple "type" and that's it (i.e., integer) etc. That will give you a number that you can pass around in infra-world and convert to a WebIDL type whenever you need to surface it to script.

Reply:
Yeah, that makes sense to me.

@JensenPaul maybe toss in a TODO here and let's land this?

Reply:
I reworked this to be a proper structured header, and I also redid the header below it likewise.